### PR TITLE
Remove `ClampZ` function from `Kokkos_Cuda_Parallel_MDRange.hpp`

### PR DESCRIPTION
**Pull request by Adrien-Tab**
### Overview
Currently, when launching a `Kokkos::parallel_for` loop with `MDRangePolicy` and a tiling size in the z direction that exceeds hardware limits, the tile is silently modified to fit the limits. see https://github.com/kokkos/kokkos/issues/8239 

The new behavior when z limit is exceeded will be the same as when the x or y limit is exceeded. see : https://github.com/kokkos/kokkos/pull/8189

### PR Changes:
- Remove clampZ function in `Kokkos_CUDA_Parallel_MDRange.hpp`.
- Constrain tiles for the tuner using `TileSizeProperties` values to prevent invalid launch configurations from being explored.
- Updated `TestBuiltinTuners.cpp` to use gtest framework.
- Added tests for tile constraints validation in `TestBuiltinTuners.cpp`. 
- ~~Add block size limits for devices in `TileSizeProperties` to provide proper hardware constraints.~~

### CODE REVIEW

File: core/src/Kokkos_Tuners.hpp [hunk 1]
```diff
@@ -637,17 +637,60 @@ class RangePolicyOccupancyTuner {
...
+      }
+    }
+  }
+}
+
+template <typename T, typename Mapped>
+void constraint_tile(std::map<T, Mapped>& cont,
+                     const Kokkos::Array<int, 3>& hw_tile_limits, int rank) {
+  if (rank > 2) {
+    return;
```
Comment (masterleinad) [inline core/src/Kokkos_Tuners.hpp]: We  should relicate the logic for mapping from up to 6 dimensions to 3 dimensions here.

Comment (Adrien-Tab) [reply core/src/Kokkos_Tuners.hpp]: I’ve extended it to 6 dimensions by replicating the logic from `MDRangePolicy` used on device. It seems like there’s not much documentation about the Kokkos tuning part.

**Review by masterleinad — COMMENTED** [review]
This all looks reasonable to me although it relies on backend-specific implementations. It would be good to have a test for `apply_tiles_constraints` and a death test when the last block dimension for Cuda and HIP to be too large despite the the total block size being OK.

**Review by masterleinad — COMMENTED** [review]
Still missing a test for `apply_tiles_constraints`.

File: core/src/Kokkos_Tuners.hpp [hunk 2]
```diff
@@ -637,17 +637,88 @@ class RangePolicyOccupancyTuner {
...
+                                             int policy_rank) {
+  Kokkos::Array<T, 3> hw_tile{1, 1, 1};
+  if (policy_rank < 4) {
+    hw_tile[0] = tile[0];
+    hw_tile[1] = tile[1];
+    hw_tile[2] = tile[2];
+  } else {
+    hw_tile[0] = tile[0] * tile[1];
+    hw_tile[1] = tile[2] * tile[3];
+    hw_tile[2] = tile[4] * tile[5];
```
Comment (pzehner) [inline core/src/Kokkos_Tuners.hpp]: I'm not sure the mapping is like this for a rank larger or equal than 4. Say we have hardware tiles $t_x$, $t_y$, and $t_z$, and we have software tiles $(t_i)_{i=0}^5$:

- Rank 4: $t_x = t_0 t_1$, $t_y = t_2$, $t_z = t_3$;
- Rank 5: $t_x = t_0 t_1$, $t_y = t_2 t_3$, $t_z = t_4$;
- Rank 6: $t_x = t_0 t_1$, $t_y = t_2 t_3$, $t_z = t_4 t_5$.

Or, am I wrong?

File: core/src/Kokkos_Tuners.hpp [hunk 3]
```diff
@@ -637,17 +637,88 @@ class RangePolicyOccupancyTuner {
...
+  }
+}
+
+template <typename KeyType, typename Mapped>
+void apply_tiles_constraints(std::map<KeyType, Mapped>& cont,
+                             const Kokkos::Array<int, 3>& hw_tile_limits,
+                             int policy_rank) {
+  Kokkos::Array<int, 6> current_tile{1, 1, 1, 1, 1, 1};
+  apply_tiles_constraints(cont, hw_tile_limits, current_tile, 0, policy_rank);
+}
```
Comment (pzehner) [inline core/src/Kokkos_Tuners.hpp]: I have a hard time following your logic. I think comments would be welcome.

**Review by pzehner — COMMENTED** [review]
I think it's a good think to get rid of this block `clampZ` hidden feature, and I also think that introducing the new tiles constraint mechanism can be interesting. However, I think the two modifications are too orthogonal to be kept in a single PR, the second one not corresponding to the PR title, neither to its description overview.

I think this PR should focus on removing the `clampZ` limitation, and a new PR regarding the tiles constraint should be opened.

File: core/src/Kokkos_Tuners.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/Kokkos_Tuners.hpp]: Yes, see 
https://github.com/kokkos/kokkos/blob/3a75ea72e10c08789a6db9ba34ca669ee0688f79/core/src/Cuda/Kokkos_Cuda_Parallel_MDRange.hpp#L179-L180
https://github.com/kokkos/kokkos/blob/3a75ea72e10c08789a6db9ba34ca669ee0688f79/core/src/Cuda/Kokkos_Cuda_Parallel_MDRange.hpp#L197-L198
https://github.com/kokkos/kokkos/blob/3a75ea72e10c08789a6db9ba34ca669ee0688f79/core/src/Cuda/Kokkos_Cuda_Parallel_MDRange.hpp#L214-L216.

**Comment by masterleinad** [comment]
> I think it's a good think to get rid of this block `clampZ` hidden feature, and I also think that introducing the new tiles constraint mechanism can be interesting. However, I think the two modifications are too orthogonal to be kept in a single PR, the second one does not correspond to the PR title, neither to its description overview.

We can't just drop `clampZ` without brealing auto-tuning, though.

File: core/src/Kokkos_Tuners.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/Kokkos_Tuners.hpp]: So the logic should be
```C++
for (int i = 0; i < rank; ++i)
  hw_tile[(3*i)/rank] *= m_tile[i];
```

**Comment by pzehner** [comment]
> We can't just drop `clampZ` without brealing auto-tuning, though.

How so? If it uses a constraint in $z$, can we just lift it?

**Comment by masterleinad** [comment]
> How so? If it uses a constraint in z , can we just lift it?

See https://github.com/kokkos/kokkos/issues/8239#issuecomment-3057388340. Auto-tuning creates a bunch of tile sizes to run through but only knows about the maximum block size/tile size product but not restrictions on particular block dimensions. Clamping made sure the kernels could still be launched (even though we now realized that they would yield wrong results).

File: core/src/Kokkos_Tuners.hpp (hunk 2 above)
Comment (Adrien-Tab) [reply core/src/Kokkos_Tuners.hpp]: I keep the logic explicit because it is easier to understand.

Comment (masterleinad) [reply core/src/Kokkos_Tuners.hpp]: But it's wrong for rank 4 and 5.

Comment (Adrien-Tab) [reply core/src/Kokkos_Tuners.hpp]: Oh yes, you're right.

File: core/unit_test/tools/TestBuiltinTuners.cpp [hunk 4]
```diff
@@ -13,84 +13,154 @@
...
+
+  fill_tile(tile_configuration_space, max_total_tile_size);
+  apply_tiles_constraints(tile_configuration_space, hw_const, test_policy_rank);
+  validate_tile_sizes(tile_configuration_space, hw_const, test_policy_rank);
+}
+
+TEST(kokkosp, constraint_tiles) {
+  namespace Impl = Kokkos::Tools::Experimental::Impl;
+
+  std::array<int, 3> hw_const{128, 128, 128};
```
Comment (yasahi-hpc) [inline core/unit_test/tools/TestBuiltinTuners.cpp]: Can you add a test case for the maximum valid tile size. 
I also love to have a failure test for an invalid tile including a zero

**Review by yasahi-hpc — CHANGES_REQUESTED** [review]
LGTM. I would prefer more stressful test cases

**Review by yasahi-hpc — APPROVED** [review]
LGTM.
Previously, a tile size including zero was allowed? 
In my opinion, a tile size including zero is invalid, but if we disallow it in this PR, it changes the behavior.

**Comment by Adrien-Tab** [comment]
> LGTM. Previously, a tile size including zero was allowed? In my opinion, a tile size including zero is invalid, but if we disallow it in this PR, it changes the behavior.

I didn't check it at first because in the set of candidate tiles it is impossible to have a tile size with a zero

**Comment by yasahi-hpc** [comment]
> > LGTM. Previously, a tile size including zero was allowed? In my opinion, a tile size including zero is invalid, but if we disallow it in this PR, it changes the behavior.
> 
> I didn't check it at first because in the set of candidate tiles it is impossible to have a tile size with a zero

Can user specify a tile with a zero?
If they compute a tile size with some formula, it can include zero?

**Comment by Adrien-Tab** [comment]
> 
> Can user specify a tile with a zero? If they compute a tile size with some formula, it can include zero?

If they do that, the constructor will automatically use a value calculated from the static value.

See : https://github.com/kokkos/kokkos/blob/develop/core/src/KokkosExp_MDRangePolicy.hpp#L405-L418

**Comment by yasahi-hpc** [comment]
> > Can user specify a tile with a zero? If they compute a tile size with some formula, it can include zero?
> 
> If they do that, the constructor will automatically use a value calculated from the static value.
> 
> See : https://github.com/kokkos/kokkos/blob/develop/core/src/KokkosExp_MDRangePolicy.hpp#L405-L418

Sure. Then I am good

File: core/src/Kokkos_Tuners.hpp [hunk 5]
```diff
@@ -637,17 +637,91 @@ class RangePolicyOccupancyTuner {
...
+
+// Entry point for applying tile constraints. Filters out invalid tiles that
+// exceed hardware limits based on the policy rank.
+template <typename KeyType, typename Mapped>
+void apply_tiles_constraints(std::map<KeyType, Mapped>& cont,
+                             const std::array<int, 3>& hw_tile_limits,
+                             int policy_rank) {
+  std::array<int, 6> current_tile{1, 1, 1, 1, 1, 1};
+  apply_tiles_constraints(cont, hw_tile_limits, current_tile, 0, policy_rank);
+}
```
Comment (masterleinad) [inline core/src/Kokkos_Tuners.hpp:704]: Can you give an explicit example of wht happens here?

File: core/unit_test/tools/TestBuiltinTuners.cpp [hunk 6]
```diff
@@ -13,84 +13,161 @@
...
+    std::array<int, 6> next_tile = current_tile;
+    validate_tile_sizes(it->second, next_tile, hw_tile_limits, current_rank + 1,
+                        policy_rank);
+  }
+}
+
+// Test helper function that verifies all tiles in the configuration space
+// satisfy hardware constraints.
+template <typename KeyType, typename Mapped>
+void validate_tile_sizes(std::map<KeyType, Mapped>& cont,
```
Comment (JBludau) [inline core/unit_test/tools/TestBuiltinTuners.cpp]: hmm not sure if I like the recurring names ... why?

**Review by JBludau — COMMENTED** [review]
removing the clamp is fine afaict. The tests are hard to read

File: core/src/Kokkos_Tuners.hpp [hunk 7]
```diff
@@ -637,17 +637,91 @@ class RangePolicyOccupancyTuner {
...
+// nested map structure. If a tile exceeds limits, remove the entire branch from
+// the research space.
+template <typename KeyType, typename Mapped>
+void apply_tiles_constraints(std::map<KeyType, Mapped>& cont,
+                             const std::array<int, 3>& hw_tile_limits,
+                             std::array<int, 6>& current_tile, int current_rank,
+                             const int policy_rank) {
+  for (auto it = cont.begin(); it != cont.end();) {
+    KeyType dimension_size     = it->first;
+    current_tile[current_rank] = dimension_size;
```
Comment (masterleinad) [inline core/src/Kokkos_Tuners.hpp:685]: Why don't we reset the tile size for the current rank to 1 after iterating over that dimension?

Comment (Adrien-Tab) [reply core/src/Kokkos_Tuners.hpp:685]: At each level of recursion, a local copy is created, so each branch works on its own copy

File: core/src/Kokkos_Tuners.hpp (hunk 5 above)
Comment (Adrien-Tab) [reply core/src/Kokkos_Tuners.hpp:704]: These recursive functions are used to traverse the space of candidate tiles. (Reminder) The structure of the candidate tiles is a nesting of `std::map` with `std::vector` as the last level (leaf). The ranks correspond to the level of nesting.

Here (L721), the code starts by creating a tile that will be copied by the recursive calls. Then the recursive function is called by specifying the level (`current_rank`) being processed to keep track through the recursive calls.

<details>
<summary>Example For 6D</summary>
I have printed all the tiles. Here is the output with the `current_rank` marked in brackets on the left. The tiles that were removed have an "X" on their right.

```
[0] (1, 1, 1, 1, 1, 1)
  [1] (1, 1, 1, 1, 1, 1)
    [2] (1, 1, 1, 1, 1, 1)
      [3] (1, 1, 1, 1, 1, 1)
        [4] (1, 1, 1, 1, 1, 1)
          [5] (1, 1, 1, 1, 1, 1024) X 
          [5] (1, 1, 1, 1, 1, 512) X 
          [5] (1, 1, 1, 1, 1, 256) X 
          [5] (1, 1, 1, 1, 1, 128) X 
          [5] (1, 1, 1, 1, 1, 64)
        [4] (1, 1, 1, 1, 2, 1)
          [5] (1, 1, 1, 1, 2, 512) X 
          [5] (1, 1, 1, 1, 2, 256) X 
          [5] (1, 1, 1, 1, 2, 128) X 
          [5] (1, 1, 1, 1, 2, 64) X 
          [5] (1, 1, 1, 1, 2, 32)
        [4] (1, 1, 1, 1, 4, 1)
          [5] (1, 1, 1, 1, 4, 256) X 
          [5] (1, 1, 1, 1, 4, 128) X 
          [5] (1, 1, 1, 1, 4, 64) X 
          [5] (1, 1, 1, 1, 4, 32) X 
          [5] (1, 1, 1, 1, 4, 16)
        [4] (1, 1, 1, 1, 8, 1)
          [5] (1, 1, 1, 1, 8, 128) X 
          [5] (1, 1, 1, 1, 8, 64) X 
          [5] (1, 1, 1, 1, 8, 32) X 
          [5] (1, 1, 1, 1, 8, 16) X 
          [5] (1, 1, 1, 1, 8, 8)
        [4] (1, 1, 1, 1, 16, 1)
          [5] (1, 1, 1, 1, 16, 64) X 
          [5] (1, 1, 1, 1, 16, 32) X 
          [5] (1, 1, 1, 1, 16, 16) X 
          [5] (1, 1, 1, 1, 16, 8) X 
          [5] (1, 1, 1, 1, 16, 4)
        [4] (1, 1, 1, 1, 32, 1)
          [5] (1, 1, 1, 1, 32, 32) X 
          [5] (1, 1, 1, 1, 32, 16) X 
          [5] (1, 1, 1, 1, 32, 8) X 
          [5] (1, 1, 1, 1, 32, 4) X 
          [5] (1, 1, 1, 1, 32, 2)
        [4] (1, 1, 1, 1, 64, 1)
          [5] (1, 1, 1, 1, 64, 16) X 
          [5] (1, 1, 1, 1, 64, 8) X 
          [5] (1, 1, 1, 1, 64, 4) X 
          [5] (1, 1, 1, 1, 64, 2) X 
          [5] (1, 1, 1, 1, 64, 1)
        [4] (1, 1, 1, 1, 128, 1) X 
        [4] (1, 1, 1, 1, 256, 1) X 
        [4] (1, 1, 1, 1, 512, 1) X 
        [4] (1, 1, 1, 1, 1024, 1) X 
      [3] (1, 1, 1, 2, 1, 1)
        [4] (1, 1, 1, 2, 1, 1)
          [5] (1, 1, 1, 2, 1, 512) X 
          [5] (1, 1, 1, 2, 1, 256) X 
          [5] (1, 1, 1, 2, 1, 128) X 
          [5] (1, 1, 1, 2, 1, 64)
```
</details>

File: core/src/Kokkos_Tuners.hpp (hunk 7 above)
Comment (masterleinad) [reply core/src/Kokkos_Tuners.hpp:685]: Would you mind passing `current_tile` by value to make that clearer?

Comment (Adrien-Tab) [reply core/src/Kokkos_Tuners.hpp:685]: I'll make that change.

File: core/src/Kokkos_Tuners.hpp [hunk 8]
```diff
@@ -638,16 +638,89 @@ namespace Impl {
...
+template <typename T, size_t N>
+constexpr bool valid_tile(const std::array<T, 3> hw_tile_limits,
+                          const std::array<T, N> current_tile,
+                          int policy_rank) {
+  auto hw_tile = map_to_hw_tile(current_tile, policy_rank);
+
+  return (0 < hw_tile[0] && hw_tile[0] <= hw_tile_limits[0] && 0 < hw_tile[1] &&
+          hw_tile[1] <= hw_tile_limits[1] && 0 < hw_tile[2] &&
+          hw_tile[2] <= hw_tile_limits[2]);
+}
```
Comment (masterleinad) [inline core/src/Kokkos_Tuners.hpp:656]: These functions are only used for `T=int` and `N=6`.

**Review by masterleinad — APPROVED** [review]
Fine with me.

File: core/src/Kokkos_Tuners.hpp (hunk 8 above)
Comment (Adrien-Tab) [reply core/src/Kokkos_Tuners.hpp:656]: You're right about the current usage. I've expanded the tests to cover more general cases with `int`, `long` and ranks 2-6.

Comment (masterleinad) [reply core/src/Kokkos_Tuners.hpp:656]: I meant that you don't need template parameters if the code is only ever instantiated for `T=int` and `N=6`.

Comment (masterleinad) [reply core/src/Kokkos_Tuners.hpp:656]: Something like
```diff
diff --git a/core/src/Kokkos_Tuners.hpp b/core/src/Kokkos_Tuners.hpp
index 7751d351d..e84c1d459 100644
--- a/core/src/Kokkos_Tuners.hpp
+++ b/core/src/Kokkos_Tuners.hpp
@@ -636,34 +636,31 @@ class RangePolicyOccupancyTuner {
 
 namespace Impl {
 
-template <typename T>
-void fill_tile(std::vector<T>& cont, int tile_size) {
+void inline fill_tile(std::vector<int>& cont, int tile_size) {
   for (int x = 1; x <= tile_size; x *= 2) {
     cont.push_back(x);
   }
 }
-template <typename T, typename Mapped>
-void fill_tile(std::map<T, Mapped>& cont, int tile_size) {
+template <typename Mapped>
+void fill_tile(std::map<int, Mapped>& cont, int tile_size) {
   for (int x = 1; x <= tile_size; x *= 2) {
     fill_tile(cont[x], tile_size / x);
   }
 }
 
 // Map MDRangePolicy tile dimensions to hardware tile dimensions
-template <typename T, size_t N>
-constexpr std::array<T, 3> map_to_hw_tile(const std::array<T, N> tile,
-                                          int policy_rank) {
-  std::array<T, 3> hw_tile{1, 1, 1};
-  for (int i = 0; i < static_cast<int>(N) && i < policy_rank; ++i) {
+constexpr std::array<int, 3> map_to_hw_tile(const std::array<int, 6> tile,
+                                            int policy_rank) {
+  std::array<int, 3> hw_tile{1, 1, 1};
+  for (int i = 0; i < 6 && i < policy_rank; ++i) {
     hw_tile[(i * 3) / policy_rank] *= tile[i];
   }
   return hw_tile;
 }
 
-template <typename T, size_t N>
-constexpr bool valid_tile(const std::array<T, 3> hw_tile_limits,
-                          const std::array<T, N> current_tile,
-                          int policy_rank) {
+constexpr inline bool valid_tile(const std::array<int, 3> hw_tile_limits,
+                                 const std::array<int, 6> current_tile,
+                                 int policy_rank) {
   auto hw_tile = map_to_hw_tile(current_tile, policy_rank);
 
   return (0 < hw_tile[0] && hw_tile[0] <= hw_tile_limits[0] && 0 < hw_tile[1] &&
@@ -673,11 +670,10 @@ constexpr bool valid_tile(const std::array<T, 3> hw_tile_limits,
 
 // Prunes invalid tiles dimension from largest to smallest until a valid size is
 // found
-template <typename T>
-void apply_tiles_constraints(std::vector<T>& cont,
-                             const std::array<int, 3>& hw_tile_limits,
-                             std::array<int, 6> current_tile, int current_rank,
-                             const int policy_rank) {
+inline void apply_tiles_constraints(std::vector<int>& cont,
+                                    const std::array<int, 3>& hw_tile_limits,
+                                    std::array<int, 6> current_tile,
+                                    int current_rank, const int policy_rank) {
   auto it = cont.end();
   while (it != cont.begin()) {
     --it;
@@ -693,13 +689,13 @@ void apply_tiles_constraints(std::vector<T>& cont,
 // Validate the current tile size and recursively apply constraints to the
 // nested map structure. If a tile exceeds limits, remove the entire branch from
 // the research space.
-template <typename KeyType, typename Mapped>
-void apply_tiles_constraints(std::map<KeyType, Mapped>& cont,
+template <typename Mapped>
+void apply_tiles_constraints(std::map<int, Mapped>& cont,
                              const std::array<int, 3>& hw_tile_limits,
                              std::array<int, 6> current_tile, int current_rank,
                              const int policy_rank) {
   for (auto it = cont.begin(); it != cont.end();) {
-    KeyType dimension_size     = it->first;
+    int dimension_size         = it->first;
     current_tile[current_rank] = dimension_size;
     if (!valid_tile(hw_tile_limits, current_tile, policy_rank)) {
       it = cont.erase(it);
@@ -713,8 +709,8 @@ void apply_tiles_constraints(std::map<KeyType, Mapped>& cont,
 
 // Entry point for applying tile constraints. Filters out invalid tiles that
 // exceed hardware limits based on the policy rank.
-template <typename KeyType, typename Mapped>
-void apply_tiles_constraints(std::map<KeyType, Mapped>& cont,
+template <typename Mapped>
+void apply_tiles_constraints(std::map<int, Mapped>& cont,
                              const std::array<int, 3>& hw_tile_limits,
                              int policy_rank) {
   std::array<int, 6> current_tile{1, 1, 1, 1, 1, 1};
diff --git a/core/unit_test/tools/TestBuiltinTuners.cpp b/core/unit_test/tools/TestBuiltinTuners.cpp
index 154247b97..66664e6cc 100644
--- a/core/unit_test/tools/TestBuiltinTuners.cpp
+++ b/core/unit_test/tools/TestBuiltinTuners.cpp
@@ -152,13 +152,14 @@ TEST(kokkosp, constraint_tiles) {
 
   std::array<int, 3> hw_const{128, 128, 128};
   std::array<int, 6> valid_tile{4, 2, 4, 4, 2, 4};
-  std::array<int, 3> valid_tile_max{128, 128, 128};
-  std::array<int, 3> tile_with_zero{128, 0, 1};
-  std::array<int, 3> invalid_tile{256, 1, 1};
+  std::array<int, 6> valid_tile_max{128, 128, 128, 1, 1};
+  std::array<int, 6> tile_with_zero{128, 0, 1, 1, 1, 1};
+  std::array<int, 6> invalid_tile{256, 1, 1, 1, 1, 1};
 
   // Impl::valid_tile checks each dimension but not the total product
   EXPECT_TRUE(Impl::valid_tile(hw_const, valid_tile, 6));
   EXPECT_TRUE(Impl::valid_tile(hw_const, valid_tile_max, 3));
+  EXPECT_FALSE(Impl::valid_tile(hw_const, valid_tile_max, 4));
   EXPECT_FALSE(Impl::valid_tile(hw_const, tile_with_zero, 3));
   EXPECT_FALSE(Impl::valid_tile(hw_const, invalid_tile, 3));
```
also works.

**Review by masterleinad — APPROVED** [review]
```
/var/jenkins/workspace/Kokkos_PR-8264/core/src/Cuda/Kokkos_Cuda_Parallel_MDRange.hpp:104:16: error: unused variable 'max_threads_dim' [clang-diagnostic-unused-variable]

    const auto max_threads_dim = m_rp.space().cuda_device_prop().maxThreadsDim;
```
Otherwise, this looks good to me. Approving already to enable SNL CI.

File: core/src/Kokkos_Tuners.hpp (hunk 8 above)
Comment (Adrien-Tab) [reply core/src/Kokkos_Tuners.hpp:656]: Applied the suggested changes, thanks.

File: core/unit_test/tools/TestBuiltinTuners.cpp [hunk 9]
```diff
@@ -13,84 +13,162 @@
...
+  validate_tile_sizes(tile_configuration_space, hardware_limits,
+                      test_policy_rank);
+}
+
+TEST(kokkosp, constraint_tiles) {
+  namespace Impl = Kokkos::Tools::Experimental::Impl;
+
+  std::array<int, 3> hardware_limits{128, 128, 128};
+  std::array<int, 6> valid_tile{4, 2, 4, 4, 2, 4};
+  std::array<int, 6> valid_tile_max{128, 128, 128, 1, 1};
```
Comment (dalg24) [inline core/unit_test/tools/TestBuiltinTuners.cpp]: Did you mean to only provide 5 numbers (i.e. implicitly zero-initialize the 6th dimension)

File: core/unit_test/tools/TestBuiltinTuners.cpp [hunk 10]
```diff
@@ -13,84 +13,162 @@
...
+  std::array<int, 6> tile_with_zero{128, 0, 1, 1, 1, 1};
+  std::array<int, 6> invalid_tile{256, 1, 1, 1, 1, 1};
+
+  // Impl::valid_tile checks each dimension but not the total product
+  EXPECT_TRUE(Impl::valid_tile(hardware_limits, valid_tile, 6));
+  EXPECT_TRUE(Impl::valid_tile(hardware_limits, valid_tile_max, 3));
+  EXPECT_FALSE(Impl::valid_tile(hardware_limits, valid_tile_max, 4));
```
Comment (dalg24) [inline core/unit_test/tools/TestBuiltinTuners.cpp]: Why do we expect `valid_tile_max` would not be a valid tile?
Is there a bug (the zero trailing size) or is the name poorly chosen?

File: core/unit_test/tools/TestBuiltinTuners.cpp [hunk 11]
```diff
@@ -13,84 +13,162 @@
...
+
+  std::array<int, 3> hardware_limits{128, 128, 128};
+  SpaceDescription tile_configuration_space;
+  const int max_total_tile_size = 512;
+
+  fill_tile(tile_configuration_space, max_total_tile_size);
+  apply_tiles_constraints(tile_configuration_space, hardware_limits,
+                          test_policy_rank);
+  validate_tile_sizes(tile_configuration_space, hardware_limits,
+                      test_policy_rank);
```
Comment (dalg24) [inline core/unit_test/tools/TestBuiltinTuners.cpp]: I find it quite confusing which are Kokkos implementation details (`fill_tile`,` apply_tiles_constraints`) and which are helpers defined in that test source file (`validate_tile_sizes`).
I suppose one option would be to use an alias namespace as you do below and qualify the implementation details ones.

File: core/unit_test/tools/TestBuiltinTuners.cpp [hunk 12]
```diff
@@ -13,84 +13,162 @@
...
   }
-  Kokkos::finalize();
 }
+
+template <typename Mapped>
+void validate_tile_sizes(std::map<int, Mapped>& cont,
+                         std::array<int, 6> current_tile,
+                         const std::array<int, 3>& hw_tile_limits,
+                         int current_rank, const int policy_rank) {
+  for (auto it = cont.begin(); it != cont.end(); it++) {
```
Comment (dalg24) [inline core/unit_test/tools/TestBuiltinTuners.cpp]:
```suggestion
  for (const auto& [key, vector] : cont) {
```

File: core/unit_test/tools/TestBuiltinTuners.cpp [hunk 13]
```diff
@@ -13,84 +13,162 @@
...
+
+// Test helper function that verifies all tiles in the configuration space
+// satisfy hardware constraints.
+template <typename Mapped>
+void validate_tile_sizes(std::map<int, Mapped>& cont,
+                         const std::array<int, 3>& hw_tile_limits,
+                         int policy_rank) {
+  std::array<int, 6> current_tile{1, 1, 1, 1, 1, 1};
+  validate_tile_sizes(cont, current_tile, hw_tile_limits, 0, policy_rank);
+}
```
Comment (dalg24) [inline core/unit_test/tools/TestBuiltinTuners.cpp:122]: This is way too confusing

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: core/unit_test/tools/TestBuiltinTuners.cpp (hunk 9 above)
Comment (Adrien-Tab) [reply core/unit_test/tools/TestBuiltinTuners.cpp]: No, it was a mistake. I corrected it.

File: core/unit_test/tools/TestBuiltinTuners.cpp (hunk 13 above)
Comment (Adrien-Tab) [reply core/unit_test/tools/TestBuiltinTuners.cpp:122]: I added comments and changed a function name

File: core/unit_test/tools/TestBuiltinTuners.cpp (hunk 11 above)
Comment (Adrien-Tab) [reply core/unit_test/tools/TestBuiltinTuners.cpp]: I'm now using a namespace alias

File: core/unit_test/tools/TestBuiltinTuners.cpp (hunk 10 above)
Comment (Adrien-Tab) [reply core/unit_test/tools/TestBuiltinTuners.cpp]: I clarified this section

File: core/unit_test/tools/TestBuiltinTuners.cpp [hunk 14]
```diff
@@ -1,83 +1,173 @@
...
+   * above. The At 0.0, this will be the first element (1).
+   * At 0.9 this will be the last element (5)
+   */
+  auto begin_point = new_team_tuner.get_point(0.0, 0.0, 0.0);
+  EXPECT_EQ(std::get<0>(begin_point), 1);
+  auto end_point = new_team_tuner.get_point(0.9, 0.0, 0.0);
+  EXPECT_EQ(std::get<0>(end_point), 5);
+  for (int x = 0; x < 10000; ++x) {
+    auto config                 = new_team_tuner.begin();
+    [[maybe_unused]] int option = std::get<0>(config);
```
Comment (dalg24) [inline core/unit_test/tools/TestBuiltinTuners.cpp:76]: Not worth re-running) the CI but I prefer cast to void or
```suggestion
    int option = std::get<0>(config);
    std::ignore = option;
```
to the `[[maybe_unused]]` attribute

File: core/unit_test/tools/TestBuiltinTuners.cpp [hunk 15]
```diff
@@ -1,83 +1,173 @@
...
+}
+
+// Builds candidate tiles by traversing the nested map structure. Each recursive
+// level sets one dimension.
+template <typename Mapped>
+void validate_tile_sizes(const std::map<int, Mapped>& cont,
+                         std::array<int, 6> current_tile,
+                         const std::array<int, 3>& hw_tile_limits,
+                         const int current_rank, const int policy_rank) {
+  for (const auto& [dimension_size, sub_cont] : cont) {
```
Comment (dalg24) [inline core/unit_test/tools/TestBuiltinTuners.cpp:107]: Is `sub_cnt` always a vector here or it can also be a map?

Comment (Adrien-Tab) [reply core/unit_test/tools/TestBuiltinTuners.cpp:107]: yes it can be either map or vector

**Review by dalg24 — APPROVED** [review]
